src/code-components/FileInput/FileInput.tsx

```diff
@@ -0,0 +1,97 @@
+import { ChangeEvent, ReactNode, useRef } from "react";
+import { FILE_TYPES } from "./FileInput.register";
+
+interface FileInputProps {
+  onChange?(files: File[]): void;
+  types?: string[];
+  customTypes?: string[];
+  multiple?: boolean;
+  maxSize?: number;
+  className?: string;
+  children: ReactNode;
+}
+
+export function FileInput({
+  onChange,
+  types,
+  customTypes,
+  multiple,
+  maxSize,
+  className,
+  children,
+}: FileInputProps) {
+  const inputRef = useRef<HTMLInputElement>(null);
+  const currentTypes =
+    customTypes ??
+    types?.flatMap((key) => FILE_TYPES[key as keyof typeof FILE_TYPES]) ??
+    [];
+
+  async function handleFilesChange(e: ChangeEvent<HTMLInputElement>) {
+    const fileList = e.target.files;
+    if (!fileList) return;
+
+    const files = Array.from(fileList);
+    const validFiles = files.filter((file) =>
+      isFileValid(file, currentTypes, maxSize),
+    );
+
+    if (validFiles.length > 0) {
+      onChange?.(validFiles);
+    }
+
+    return;
+  }
+
+  function handleClick() {
+    inputRef.current?.click();
+  }
+
+  const accept = getAccept(currentTypes);
+
+  return (
+    <>
+      <input
+        type="file"
+        ref={inputRef}
+        accept={accept}
+        multiple={multiple}
+        onChange={handleFilesChange}
+        hidden
+      />
+      <div onClick={handleClick} className={className}>
+        {children}
+      </div>
+    </>
+  );
+}
+
+function getAccept(types: string[]) {
+  if (!types.length) return undefined;
+  return types.join(",");
+}
+
+function isFileValid(file: File, types: string[], maxSize?: number) {
```

Comment:
This should be cleaner and more accurate I think?

```suggestion
function isFileValid(file: File, allowedFormats: string[], maxSize?: number) {
```